### DISCUSSION TITLE
How do you handle popups in Tauri?

### DISCUSSION THREAD

**Discussion by elanzini** [tauri-apps/tauri#5251] [Q&A]
I am trying to implement [Google OAuth 2.0 flow](https://developers.google.com/identity/protocols/oauth2/javascript-implicit-flow) from my Tauri application. I am getting stuck because the application does not allow popups  - `error:"popup_blocked_by_browser"`.

1. Does Tauri by design not allow popups? 
2. Is there a way to handle them in the framework?
3. More specifically to OAuth 2.0, what would be the `redirect_uri` to use to redirect to the desktop application?

**Comment by FabianLars** [comment]
popups won't be actual tauri windows, but something the webviews handle internally. At least WebView2 (Windows) can handle popups in theory, but it seems to be a little random when it opens them and when not - or maybe it depends on how the popup is request or whatever 🤔 

Another problem is that google forbids oauth in non-browser environments, so even after the popup opens, it should show a on-screen warning. For what it's worth you can prevent that warning by changing the user agent - at least until google figures out a better way to filter out webviews.

redirect_uris are tricky too, realistically you can't redirect to the actual tauri app (`tauri://` and `https://tauri.localhost/`), primarily because it will break the access to Tauri APIs and i think google doesn't allow non-http schemes either but i could be wrong.
I don't know if there are workarounds, but from what i've seen so far you'll need a localhost listener at which point you may as well follow the recommendation and let the user log in in their default browser 🤷

**Reply by elanzini** [reply]
Thanks for the detailed answer.
Can you link some resources regarding the localhost listener? How do you set it up in Tauri?

Are you referring to the `tauri-plugin-localhost`?

**Reply by FabianLars** [reply]
> Can you link some resources regarding the localhost listener? How do you set it up in Tauri?

Well, there is no default thing for that yet, everybody does their own thing currently. I even started exploring a minimal approach myself here: https://github.com/FabianLars/tauri-plugin-oauth - take the word `plugin` with a grain of salt, there is no js api yet. Speaking of js, starting a localhost server of any kind will require rust (the idea of plugins is to abstract this away for the user).

> Are you referring to the tauri-plugin-localhost?

No this plugin hosts your whole frontend / all your assets on a localhost server which obviously should only be used when absolutely needed.

**Reply by elanzini** [reply]
About this: `you may as well follow the recommendation and let the user log in in their default browser`
Assuming you handle the auth in the default browser, how can you pass the access_token to the Tauri application?

Deeplinks seem to be the solution but not supported atm, correct?

**Reply by FabianLars** [reply]
> Deeplinks seem to be the solution but not supported atm, correct?

This or a localhost server. In the context of _google_ oauth deep linking isn't support at all tho (by google) and requires a localhost server (which the plugin i linked above does, because it was designed with "sign in with google" in mind.

**Reply by elanzini** [reply]
Sorry I don't think I was clear enough. What I meant in the message above is that you can have a separate website (mytauriapp.com) where the user signs up (with google oauth) and there you can implement the oauth2.0 flow without issues since it's from a default browser. At this point, you have an `access_token` and a `refresh_token` - from oath2.0 and you want to pass those to the tauri app running on the same computer.

I mentioned deep links because I thought you could pass them to the Tauri app like this: `mytauriapp://signup?access_token=X&refresh_token=Y`. But perhaps at this point, you might as well have the user create an account in the website and then simply sign in in the tauri app locally.

Does it make sense? Am I going crazy?

**Reply by FabianLars** [reply]
i did understand you correctly. The localhost server i'm talking about would be your `redirect_uri` and since the localhost server is part of the tauri app you can get the tokens into your app somewhat easily.

But if you have a completely different website which handles the login and redirect_uri then not much would actually change, you'd still use the localhost server, or in this case deep links could work too, to transport it to the app.

**Reply by elanzini** [reply]
But we agreed above that non-http urls are not supported as `redirect_uri` by google so you would need to handle the entire oauth2.0 flow with mytauriapp.com. Then, once you have the tokens, you need a way to pass them along to the tauri application locally. And based on my understanding, you can only do that with deeplinks, which are not supported (yet) - correct?

**Reply by FabianLars** [reply]
a localhost server would be behind a http url tho, like `http://localhost` or `http://127.0.0.1`. And to be clear, some people use the oauth plugin from above, or something similar, successfully with google (and some other sites that i can't remember) already.

**Reply by elanzini** [reply]
> some people use the oauth plugin from above, or something similar, successfully with google (and some other sites that i can't remember) already.

will give that a try following the example in the repository. thanks a lot

**Reply by FabianLars** [reply]
The example is empty and i forgot to remove it again :/

**Edit:** I added the rust side of the example at least. The frontend is still empty tho.

The gist of it (copied from [discord](https://discord.com/channels/616186924390023171/993157913994608660/1003671353850875964)) is:

You need to specify http://localhost or http://127.0.0.1 without a port as the redirect_uri in their settings.
Once you're ready to start the server, you call the start function and give it a function it would call after the request (the plugin gives the url as an arg), the function could be something like this
```
|url| {
  // Note: You really really should validate the url on the rust side because the server is open to the whole system.
  app_handle.emit("url", url);
}
```
start() will return the port it's running on which you need for the redirect_uri setting in the oauth request thingy.

**Reply by elanzini** [reply]
I tried to integrate it following the example in https://github.com/FabianLars/tauri-plugin-oauth/commit/0c354f5aefa93b1703f7757b36919f0a4a16ff67 but I am incurring in the following errors:

<img width="1028" alt="Screen Shot 2022-09-22 at 22 54 50" src="https://user-images.githubusercontent.com/37532050/191858373-9572f379-5cf3-4186-8043-2f525adf9e07.png">

**Reply by FabianLars** [reply]
Do you have a different `Result` type imported or defined?

**Reply by elanzini** [reply]
I am just importing `use tauri::*;` above and then there is no other import/definition in the file.

**Reply by FabianLars** [reply]
Yeah no, can't recommend `*` imports at all in Rust.
The problem in this case is that you're importing tauri::Result which overwrites the std Result type.

If you wanna keep using this kind of imports then you could make the function return value more specific: `std::result::Result`

**Reply by elanzini** [reply]
That was indeed the solution. Thanks a lot. Sorry for the noob questions

**Reply by elanzini** [reply]
To piece things together, essentially, instead of opening the pop-up, I need to override the default behavior and open the link in a new tab, for example using the [shell api](https://tauri.app/v1/api/js/shell/#open). Then the user will click on consent and at this point, the `redirect_uri` will point back to `localhost`, which will be intercepted by the Tauri application running locally. 

What if you have other things running locally on different ports?
Is it better to handle the exchange of tokens and then pass it to the application with deeplinks?

**Reply by FabianLars** [reply]
> To piece things together, essentially, instead of opening the pop-up, I need to override the default behavior and open the link in a new tab, for example using the [shell api](https://tauri.app/v1/api/js/shell/#open). Then the user will click on consent and at this point, the redirect_uri will point back to localhost, which will be intercepted by the Tauri application running locally.

Yes.

> What if you have other things running locally on different ports?

Do you mean in context of port conflicts? Nothing as the plugin chooses a random available port. Security-wise? Of course a little tricky, which is why my comments in the plugin were so harsh on validating the input. Just have in mind that the whole system can send stuff to the localhost server if it finds the point.
Edit: Ah you mean that the oauth redirect chooses the wrong application? That can't happen either, because you specify the port in the actual request, just not in the google settings.

> Is it better to handle the exchange of tokens and then pass it to the application with deeplinks?

deeplinks basically have the same security complications, everything on the same system can send stuff to your app via deeplinks.

**Comment by igorjacauna** [comment]
Hi folks, just came here to say that I created [this](https://github.com/igorjacauna/tauri-firebase-login) simple project that makes login on Firebase through Google OAuth 2.0. Using [OAuth Tauri Plugin](https://github.com/FabianLars/tauri-plugin-oauth) to help the process
